webapi/Services/MemoryMigration/ChatMemoryMigrationService.cs

```diff
@@ -0,0 +1,146 @@
+using CopilotChat.WebApi.Storage;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Microsoft.SemanticKernel.Memory;
+using Microsoft.SemanticMemory;
+
+namespace CopilotChat.WebApi.Services.MemoryMigration;
+
+/// <summary>
+/// Service implementation of <see cref="IChatMemoryMigrationService"/>.
+/// </summary>
+public class ChatMemoryMigrationService : IChatMemoryMigrationService
+{
+    private readonly ILogger<ChatMemoryMigrationService> _logger;
+    private readonly ISemanticTextMemory _memory;
+    private readonly ISemanticMemoryClient _memoryClient;
+    private readonly ChatSessionRepository _chatSessionRepository;
+    private readonly ChatMemorySourceRepository _memorySourceRepository;
+    private readonly string _globalIndex;
+    private readonly PromptsOptions _promptOptions;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="ChatMigrationMonitor"/> class.
+    /// </summary>
+    public ChatMemoryMigrationService(
+        ILogger<ChatMemoryMigrationService> logger,
+        IOptions<DocumentMemoryOptions> documentMemoryOptions,
+        IOptions<PromptsOptions> promptOptions,
+        ISemanticMemoryClient memoryClient,
+        ChatSessionRepository chatSessionRepository,
+        ChatMemorySourceRepository memorySourceRepository,
+        SemanticKernelProvider provider)
+    {
+        this._logger = logger;
+        this._promptOptions = promptOptions.Value;
+        this._memoryClient = memoryClient;
+        this._chatSessionRepository = chatSessionRepository;
+        this._memorySourceRepository = memorySourceRepository;
+        this._globalIndex = documentMemoryOptions.Value.GlobalDocumentCollectionName;
+        var kernel = provider.GetMigrationKernel();
+        this._memory = kernel.Memory;
+    }
+
+    ///<inheritdoc/>
+    public async Task MigrateAsync(CancellationToken cancellationToken = default)
+    {
+        try
+        {
+            await this.InternalMigrateAsync(cancellationToken).ConfigureAwait(false);
+        }
+        catch (Exception exception) when (!exception.IsCriticalException())
+        {
+            this._logger.LogError(exception, "Error migrating chat memories");
+        }
+    }
+
+    private async Task InternalMigrateAsync(CancellationToken cancellationToken = default)
+    {
+        var collectionNames = (await this._memory.GetCollectionsAsync(cancellationToken).ConfigureAwait(false)).ToHashSet(StringComparer.OrdinalIgnoreCase);
+
+        var tokenMemory = await GetTokenMemory(cancellationToken).ConfigureAwait(false);
+        if (tokenMemory != null)
+        {
+            //  Create memory token already exists
+            return;
+        }
+
+        //  Create memory token
+        var token = Guid.NewGuid().ToString();
+        await SetTokenMemory(token, cancellationToken).ConfigureAwait(false);
+
+        await RemoveMemorySourcesAsync().ConfigureAwait(false);
+
+        bool needsZombie = true;
```

Comment:
A comment describing the role of  `needsZombie` would be helpful.

Reply:
No doubt...good point.

Reply:
Updated